pkg/alertmanager/multitenant.go

```diff
@@ -629,20 +632,27 @@ func (am *MultitenantAlertmanager) syncConfigs(cfgs map[string]alertspb.AlertCon
 		am.multitenantMetrics.lastReloadSuccessfulTimestamp.WithLabelValues(user).SetToCurrentTime()
 	}
 
+	userAlertmanagersToStop := map[string]*Alertmanager{}
+
 	am.alertmanagersMtx.Lock()
-	defer am.alertmanagersMtx.Unlock()
 	for userID, userAM := range am.alertmanagers {
 		if _, exists := cfgs[userID]; !exists {
-			level.Info(am.logger).Log("msg", "deactivating per-tenant alertmanager", "user", userID)
-			userAM.Stop()
+			userAlertmanagersToStop[userID] = userAM
 			delete(am.alertmanagers, userID)
 			delete(am.cfgs, userID)
 			am.multitenantMetrics.lastReloadSuccessful.DeleteLabelValues(userID)
 			am.multitenantMetrics.lastReloadSuccessfulTimestamp.DeleteLabelValues(userID)
 			am.alertmanagerMetrics.removeUserRegistry(userID)
-			level.Info(am.logger).Log("msg", "deactivated per-tenant alertmanager", "user", userID)
 		}
 	}
+	am.alertmanagersMtx.Unlock()
+
+	// Now stop alertmanagers and wait until they are really stopped, without holding lock.
+	for userID, userAM := range userAlertmanagersToStop {
+		level.Info(am.logger).Log("msg", "deactivating per-tenant alertmanager", "user", userID)
+		userAM.StopAndWait()
```

Comment:
What if a new alertmanager (for the same user) will be created in the meanwhile? An AM instance creation can be triggered via HTTP (see `serveRequest`  > `alertmanagerFromFallbackConfig` > `setConfig`).

Reply:
That will be a different instance. Here we stop old instance.

Reply:
Isn't there a chance we would call `StopAndWait` twice? Is that safe?

Reply:
> Isn't there a chance we would call `StopAndWait` twice? Is that safe?

I don’t see how that could happen. We remove instance from the map first (with lock held), and stop this removed instance. If new instance is added in before or while stopping previous one, it is completely separate instance, and it won’t be stopped.